Add Platybot blog post: How We Built an AI-Powered Analytics Assistant (#17465)
pulumi/docs: content/blog/how-we-built-platybot-an-ai-powered-analytics-assistant/index.md

## The problem with throwing AI at your Data Warehouse

The naive solution is obvious: connect an LLM to your database and let it write SQL. But this fails in practice, and the failure mode is insidious. Consider a few examples from our warehouse:

- **ARR is a snapshot metric.** If you query ARR (Annual Recurring Revenue) without filtering by end-of-period dates (last day of month or quarter), you get duplicate rows and wildly inflated numbers. An LLM doesn't automatically know this.
- **Account queries need exclusions.** Most queries should exclude Pulumi's own internal accounts and deleted ones. Without these filters, you're counting test data alongside real customers.
- **User queries need employee filters.** Querying active users without excluding Pulumi employees inflates adoption metrics.

The risk isn't that the AI fails visibly. It's that it succeeds convincingly with wrong numbers. A confidently wrong ARR figure presented to leadership is worse than no answer at all.

> This isn't unique to Pulumi. Every growing company hits this wall. Dashboards answer yesterday's questions, not today's. Ad-hoc LLM queries answer today's questions, but incorrectly. The gap between "I have a question" and "I have a trustworthy answer" is where data teams can get stuck.

Reviewer: ```suggestion
Many organizations run into the same constraint once data usage spreads. Dashboards answer yesterday's questions, not today's. Ad-hoc LLM queries answer today's questions, but incorrectly. The gap between "I have a question" and "I have a trustworthy answer" is where data teams can get stuck.
```

```mermaid
flowchart TB
    subgraph Entry["Entry points"]
        Web["Web UI"]
        Slack["Slack (@platybot)"]
        MCP["MCP Server"]
    end

    subgraph Core["Platybot core"]
        BE["Backend (Express + LLM)"]
    end

    subgraph Data["Data layer"]
        Cube["Cube semantic layer"]
        SF["Snowflake"]
    end

    Web --> BE
    Slack --> BE
    MCP --> BE
    BE -->|"Structured queries"| Cube
    BE -->|"Fallback SQL (read-only)"| SF
    Cube --> SF
```

The workflow is: discover domains, explore cubes, understand the schema, construct and execute the query. This mirrors how a data analyst would work. You don't jump straight to SQL; you first understand what data is available and what the metrics mean.

The AI's role is narrower than people expect. It's a translator between human intent and a well-defined data model, not a general-purpose data scientist. This is a feature, not a limitation. By constraining the AI to operate within the semantic layer, we get reliability. The creativity goes into understanding the question, not into inventing SQL.

## Meeting users where they are

> Building a good AI backend is only half the problem. If people don't use it, it doesn't matter. We launched Platybot across three interfaces, each designed for a different workflow.

Reviewer: Optional suggestion: Just removing "AI" again, for more variety of word usage. 
```suggestion
The backend solved query generation. It did not solve usage and if people don't use it, it doesn't matter. We launched Platybot across three interfaces, each designed for a different workflow.
```

### The web UI

The web app is the primary interface: a React 19 + TypeScript + Vite + Tailwind conversational UI where employees can explore data through multi-turn conversations. It supports table visualization, data export, and conversation history so you can pick up where you left off.

> Every analysis produces a shareable report with a permanent link, protected behind company authentication. The report shows the AI's reasoning so you can verify its approach, and lists every query and table used along with a sample of the data. Each query includes a direct link to run it in Metabase, our reporting tool, so any useful query can be saved, scheduled, or extended without starting from scratch.

Reviewer: Again, optionally saying AI a bit less.
```suggestion
Every analysis produces a shareable report with a permanent link, protected behind company authentication. The report shows the agent's reasoning so you can verify its approach, and lists every query and table used along with a sample of the data. Each query includes a direct link to run it in Metabase, our reporting tool, so any useful query can be saved, scheduled, or extended without starting from scratch.
```

                targetGroup: targetGroup,
            }],
            environment: [
                { name: "CUBE_API_URL", value: cubeApiUrl },
                { name: "NODE_ENV", value: "production" },
            ],
            secrets: [
                { name: "ANTHROPIC_API_KEY", valueFrom: anthropicKeySecret.arn },
                { name: "SNOWFLAKE_PASSWORD", valueFrom: snowflakePasswordSecret.arn },
            ],
            logConfiguration: {
                logDriver: "awslogs",
                options: {
                    "awslogs-group": logGroup.name,
                    "awslogs-region": aws.config.region!,
                    "awslogs-stream-prefix": "platybot",
                },
            },
        },
    },
});
```

Dogfooding Pulumi for our internal tools has real benefits beyond the obvious. Staging environments are trivial: spin up a full copy of the stack with different parameters. Secrets management is built in, so API keys and database credentials never touch a config file. And when something needs to change, the diff-and-preview workflow (`pulumi preview`) catches mistakes before they hit production.

## Results

> Since launch in September 2025: over 1,700 questions from 83 employees across every team. Usage grew steadily — from around 8 questions per day in the first month to 18 per day by January 2026, with 51 unique users that month alone. Not tire-kicking — real work. Customer analysis for sales calls, resource breakdowns for account managers, blog performance metrics for marketing, policy adoption research for product, ARR deep-dives for leadership.

Reviewer: It reads fine, but just dropping some emdashes can make it seem less likel to be LLM assisted. 

```suggestion
Since launch in September 2025: over 1,700 questions from 83 employees across every team. Usage grew steadily — from around 8 questions per day in the first month to 18 per day by January 2026, with 51 unique users that month alone. This was real production work, not experimentation. Customer analysis for sales calls, resource breakdowns for account managers, blog performance metrics for marketing, policy adoption research for product, ARR deep-dives for leadership.
```

The impact on the data team was immediate. Questions that used to land in the #analytics channel and wait for a human now get answered in seconds. The data team shifted from answering routine queries to building better models and improving data quality. We went from being a help desk to being a platform team.

Accuracy is harder to quantify, but the semantic layer gives us confidence. Because Platybot uses pre-defined measures instead of generating arbitrary SQL, entire categories of errors (wrong joins, missing filters, incorrect aggregations) are structurally less likely. When the model does get something wrong, it's usually in interpreting the question, not in the data — and users can verify that through the report's reasoning trace.

## What we learned

**The semantic layer matters more than the model.** We spent more time defining metrics in Cube than we did on any AI work. That investment pays for itself: swap the model, and the answers stay correct. Swap the semantic layer, and nothing works.

> **Meet users where they already are.** A standalone app isn't enough. Slack catches the quick questions, the web UI handles deep exploration, and MCP makes data ambient for AI-native workflows. Each interface serves a different mode of thinking.

Reviewer: ```suggestion
**Meet users where they already are.** Different UIs handle different cognitive loads. Slack catches the quick questions, the web UI handles deep exploration, and MCP makes data ambient for AI-native workflows. Each interface serves a different mode of thinking.
```
